### ISSUE TITLE
global option to set default floating window border

### ISSUE THREAD

**Issue by monkoose** [neovim/neovim#20202]
### Feature already in Vim?

No

### Feature description

**Why?**
Currently unifying the look and feel of neovim ui is kind of hard to deal with because of absence of such options.

From user perspective if you want something more unique then `single` or `rounded` borders you would need for every plugin that is using floating windows at some point to add few lines of code for controlling (if plugin allows this) its look.

From plugin author perspective you need to make it public such an option (complicate plugin config) instead of something like `nvim_open_win(.... { border = vim.get_float_border() }` and forget about it.

---
Because of `:h api-contract` and some limitation of vim options (can't be real lists or dicts) it's hard to deal with such problem.

<details><summary><b>What I suggest</b></summary>
Not sure if it is the best thing (maybe community/devs can find a better one solution).
Lets introduce two new options and one new function (`vim` or `vim.api`):
- first option is `floatborder` (or you name it) which can be one of already provided options of the border for `nvim_open_win` such as *none, single, double, rounded, solid, shadow* or any others plus one new option named *borderchars*.
- second option is `borderchars` itself with is a comma-separated characters (or maybe something like *fillchars* for more conrol) for the border the same as array for `nvim_open_win()` border option.
- our new function should just from `floatborder` and `borderchars` options generate string or array required for `nvim_open_win()` border option.

So at the end users will just need to change such two options to have the same borders. And plugin authors just to call the function to generate border option when they add `nvim_open_win()` in their plugin.
</details>

Any ideas how to improve this suggestion? What do you think overall about this feature request?

---
**EDIT** as **bfredl** suggested just some lua option like `vim.ui.border_style` seems even less work to implement this and for users and plugin authors to use.

**Comment by bfredl** [comment]
Sounds overcomplicated. why not just a convention like a `vim.ui.border_style` variable. Then plugins can just `border=vim.ui.border_style or "single"` or whatever.

**Comment by monkoose** [comment]
> Sounds overcomplicated. why not just a convention like a `vim.ui.border_style` variable. Then plugins can just `border=vim.ui.border_style or "single"` or whatever.

@bfredl because you are smarter then me:) Yes, it sounds much better. I was just limited that it should obviously be and **vim** option in my thought.

**Comment by clason** [comment]
> Maybe better to make it always present with default to 'none'? Removing the complexity of adding or.

We talked about this; `none` is the documented default and hence covered by `:h api-contract`.

**Comment by monkoose** [comment]
@clason Making default of some `vim.ui.border_style` variable to any value doesn't have anything to do with contract. Or do I not understand something?

**Comment by clason** [comment]
Sorry, I misunderstood then. The variable is not covered by any contract. (But I'm not sure that this changes anything since plugins will probably need to account for users running a Neovim without this variable.)

**Comment by monkoose** [comment]
> plugins will probably need to account for users running a Neovim without this variable.

Ah, yes, thats true.

**Comment by clason** [comment]
Note that you can just use `vim.g.border_style`; then no change in Neovim is needed, plugins will just have to agree to this convention.

**Comment by monkoose** [comment]
Yes, but if it would be documented somewhere in the text of border option of `nvim_open_win()` that you could set this option and in updated plugins expect border that you set with such option. It would be just less confused for users and plugin authors.

**Comment by clason** [comment]
Well, but that would amount to Neovim prescribing how plugins should behave. But that's not how we roll -- Neovim documents its own behavior, and plugins can do with it whatever they damn well please.

**Comment by monkoose** [comment]
> Well, but that would amount to Neovim prescribing how plugins should behave. But that's not how we roll -- Neovim documents its own behavior, and plugins can do with it whatever they damn well please.

Of course that some plugin author can decide to not follow such convention (it should still in my mind be documented, so users and plugin authors will find such variable easier), but i just want to see how mad users of such plugin be mad after this convention would be accepted by the majority.

**Comment by glepnir** [comment]
If the plugin requires many floating windows. Then I think the authors of these plugins will do some encapsulation based on `nvim_open_win`. They simply pass a border or read the border from the plugin's configuration.  There are too many example plugins of doing this . like telescope lspsaga etc . implement it is not hard . but have this option  or not makes no difference to me.

**Comment by monkoose** [comment]
> They simply pass aborder or read the border from the plugin's configuration.

Exactly, but they should publicly make such on option available to the users (still a little of extra work more documentation to add etc), and from the  user expirience it for making borders look the same for every plugin should read every documentation how author called such option is it even available etc etc. And then pollute theirs configs with a lot of the same additional lines which just setting float border. Instead of setting one variable in one place.

**Comment by xuanhung1509** [comment]
I would wait for this to be implemented someday. It's not only neovim to make a global option, but it's also the story whether plugin authors respect the option or not, but once it's done the user experience will be improved a lot.

**Comment by gpanders** [comment]
Related: https://github.com/neovim/neovim/issues/25514

I'm not ready to merge this issue into #25514 just yet, as it's not yet clear that such an option will actually be part of `vim.ui` or not (see discussion in that issue).
